yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,1048 @@
+    s" bonded to ATS=$bondedToATS;" +
+    s" listening=$listening;" +
+    s" batchSize=$batchSize;" +
+    s" flush count=$getFlushCount;" +
+    s" total number queued=$eventsQueued, processed=$eventsProcessed;" +
+    s" attempted entity posts=$postAttempts" +
+    s" successful entity posts=$postSuccesses" +
+    s" failed entity posts=$postFailures;" +
+    s" events dropped=$eventsDropped;" +
+    s" app start event received=$appStartEventProcessed;" +
+    s" app end event received=$appEndEventProcessed;"
+  }
+
+  /**
+   * Is the service listening to events from the spark context?
+   * @return true if it has registered as a listener
+   */
+  def listening: Boolean = {
+    listener.isDefined
+  }
+
+  /**
+   * Is the service hooked up to an ATS server. This does not
+   * check the validity of the link, only whether or not the service
+   * has been set up to talk to ATS.
+   * @return true if the service has a timeline client
+   */
+  def bondedToATS: Boolean = {
+    _timelineClient.isDefined
+  }
+
+  /**
+   * Set the YARN binding information. This is called during startup. It is private
+   * to the package so that tests may update this data
+   * @param appId YARN application ID
+   * @param attemptId optional attempt ID
+   */
+  private[yarn] def bindToYarnApplication(appId: ApplicationId,
+      attemptId: Option[ApplicationAttemptId]): Unit = {
+    require(appId != null, "Null appId parameter")
+    _applicationId = appId
+    _attemptId = attemptId
+  }
+
+  /**
+   * Set the "spark" application and attempt information -the information
+   * provided in the start event. The attempt ID here may be `None`; even
+   * if set it may only be unique amongst the attempts of this application.
+   * That is: not unique enough to be used as the entity ID
+   * @param appId application ID
+   * @param attemptId attempt ID
+   */
+  private def setContextAppAndAttemptInfo(appId: Option[String],
+      attemptId: Option[String]): Unit = {
+    logDebug(s"Setting application ID to $appId; attempt ID to $attemptId")
+    sparkApplicationId = appId
+    sparkApplicationAttemptId = attemptId
+  }
+
+  /**
+   * Add the listener if it is not disabled.
+   * This is accessible in the same package purely for testing
+   * @return true if the register was enabled
+   */
+  private[yarn] def registerListener(): Boolean = {
+    assert(sparkContext != null, "Null context")
+    if (sparkContext.conf.getBoolean(REGISTER_LISTENER, true)) {
+      logDebug("Registering listener to spark context")
+      val l = new YarnEventListener(sparkContext, this)
+      listener = Some(l)
+      sparkContext.listenerBus.addListener(l)
+      true
+    } else {
+      false
+    }
+  }
+
+  /**
+   * Queue an action, or if the service's `stopped` flag
+   * is set, discard it
+   * @param event event to process
+   * @return true if the event was queued
+   */
+  def enqueue(event: SparkListenerEvent): Boolean = {
+    if (!stopped.get) {
```

Comment:
This doesn't really look thread-safe. This thread can be incrementing counters and calling `handleEvent` while the thread calling `stop()` might be running code in `asyncFlush()` or interrupting the event-handling thread. Are those situations safe?